File: Examples/Rust/src/examples/ping.rs
use crate::helpers;
use ximu3::connection::*;
use ximu3::ping_response::*;
use ximu3::port_scanner::*;

pub fn run() {
    // Search for connection
    let devices = PortScanner::scan_filter(PortType::Usb);

    std::thread::sleep(std::time::Duration::from_secs(1)); // wait for OS to release port

    let Some(device) = devices.first() else {
        println!("No USB connections available");
        return;
    };

    println!("Found {device}");

    // Open connection
    let connection = Connection::new(&device.connection_config);

    if let Err(error) = connection.open() {
        println!("Unable to open {}. {error}.", connection.get_config());
        return;
    }

    // Ping
    if helpers::yes_or_no("Use async implementation?") {
        let closure = Box::new(|response| {
            print_response(response);
        });

        connection.ping_async(closure);

        std::thread::sleep(std::time::Duration::from_secs(3));
    } else {
        print_response(connection.ping());
    }

    // Close connection
    connection.close();
}

fn print_response(response: Option<PingResponse>) {
    match response {
        Some(response) => {
            println!("{}, {}, {}", response.interface, response.device_name, response.serial_number);
            // println!("{response}"); // alternative to above
        }
        None => println!("No response"),
    }
}

File: Examples/C/examples/ping.c
#include "../helpers.h"
#include <stdio.h>
#include "Ximu3.h"

static void callback(const XIMU3_PingResponse response, void *context);

static void print_response(const XIMU3_PingResponse response);

void ping() {
    // Search for connection
    const XIMU3_Devices devices = XIMU3_port_scanner_scan_filter(XIMU3_PortTypeUsb);

    sleep(1); // wait for OS to release port

    if (devices.length == 0) {
        printf("No USB connections available\n");
        return;
    }

    printf("Found %s\n", XIMU3_device_to_string(devices.array[0]));

    // Open connection
    XIMU3_Connection *const connection = XIMU3_connection_new_usb(devices.array[0].usb_connection_config);

    XIMU3_devices_free(devices);

    const XIMU3_Result result = XIMU3_connection_open(connection);

    if (result != XIMU3_ResultOk) {
        printf("Unable to open %s. %s.\n", XIMU3_connection_get_config_string(connection), XIMU3_result_to_string(result));
        XIMU3_connection_free(connection);
        return;
    }

    // Ping
    if (yes_or_no("Use async implementation?")) {
        XIMU3_connection_ping_async(connection, callback, NULL);

        sleep(3);
    } else {
        print_response(XIMU3_connection_ping(connection));
    }

    // Close connection
    XIMU3_connection_close(connection);
    XIMU3_connection_free(connection);
}

static void callback(const XIMU3_PingResponse response, void *context) {
    print_response(response);
}

static void print_response(const XIMU3_PingResponse response) {
    if (strlen(response.interface) == 0) {
        printf("No response\n");
        return;
    }

    printf("%s, %s, %s\n", response.interface, response.device_name, response.serial_number);
    // printf("%s\n", XIMU3_ping_response_to_string(response)); // alternative to above
}
